stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -2,30 +2,233 @@
 
 package com.facebook.stetho.inspector.protocol.module;
 
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.elements.DOMProvider;
+import com.facebook.stetho.inspector.elements.NodeAttribute;
+import com.facebook.stetho.inspector.elements.NodeDescriptor;
+import com.facebook.stetho.inspector.elements.NodeType;
+import com.facebook.stetho.inspector.helper.ChromePeerManager;
+import com.facebook.stetho.inspector.helper.ObjectIdMapper;
+import com.facebook.stetho.inspector.helper.PeersRegisteredListener;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcPeer;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcResult;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsDomain;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsMethod;
+import com.facebook.stetho.json.ObjectMapper;
 import com.facebook.stetho.json.annotation.JsonProperty;
 
 import org.json.JSONObject;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
 public class DOM implements ChromeDevtoolsDomain {
-  public DOM() {
+  private final ChromePeerManager mPeerManager;
+  private final DOMProvider.Factory mDOMProviderFactory;
+
+  private ObjectMapper mObjectMapper;
+  private DOMProvider mDOMProvider;
+  private DOMObjectIdMapper mObjectIdMapper;
+
+  public DOM(DOMProvider.Factory providerFactory) {
+    mDOMProviderFactory = Util.throwIfNull(providerFactory);
+
+    mPeerManager = new ChromePeerManager();
+    mPeerManager.setListener(new PeerManagerListener());
+  }
+
+  private final class PeerManagerListener extends PeersRegisteredListener {
+    @Override
+    protected void onFirstPeerRegistered() {
+      assert(mDOMProvider == null);
+
+      mObjectMapper = new ObjectMapper();
```

Comment:
This is unnecessary FWIW.  `ObjectMapper` is basically just static functions (we should have made them static functions actually).  It has no initialization and accounts for just 1 object.  Likewise, `mObjectIdMapper` is also very small if you clear the data structure.  It's only `mDOMProvider` that's considered large and is special cased using this pattern.

It's your choice as to whether you adjust things, but you must use `@Nullable` for all fields that you deinitialization in this way which is where I think you'll find that the choice is a little uglier than I think you'd prefer (IntelliJ will start warning you all over about possible null access).